proto/prostgen/rules.bzl

```diff
@@ -0,0 +1,325 @@
+"""
+A set of rules and macros for generating prost-based rust protobuf code.
+"""
+
+load("@rules_proto//proto:defs.bzl", "ProtoInfo")
+load("@io_bazel_rules_rust//rust:rust.bzl", "rust_library")
+
+_PROST_COMPILE_DEPS = [
+    "//proto/prostgen/raze:prost",
+    "//proto/prostgen/raze:prost_types",
+]
+
+_PROST_COMPILE_PROC_MACRO_DEPS = [
+    "//proto/prostgen/raze:prost_derive",
+]
+
+_TONIC_COMPILE_DEPS = [
+    "//proto/prostgen/raze:tonic",
+]
+
+ProstGenInfo = provider(
+    fields = {
+        "crate": "name of the crate that is used by the rust_library",
+        "descriptor_set": "the file descriptor set for the target's sources",
+        "transitive_proto_path": "depset containing all of the include paths " +
+                                 "that need to be passed to protoc",
+        "transitive_sources": "depset containing all of the proto files that " +
+                              "must be accessed by protoc",
+        "transitive_externs": "depset containing all of the extern targets " +
+                              "that are passed to prostgen",
+    },
+)
+
+def _external_flag(f):
+    return ["--external=%s,%s" % (
+        f[ProstGenInfo].crate,
+        f[ProstGenInfo].descriptor_set.path,
+    )]
+
+def _prost_generator_impl(ctx):
+    output_dir = ctx.actions.declare_directory(ctx.label.name)
+    lib_rs = ctx.actions.declare_file("%s/lib.rs" % ctx.label.name)
+    args = ctx.actions.args()
+
+    direct_sources = depset(transitive = [
+        depset(f[ProtoInfo].direct_sources)
+        for f in ctx.attr.deps
+    ])
+
+    transitive_sources = depset(transitive = [
+        f[ProstGenInfo].transitive_sources
+        for f in ctx.attr.extern
+    ] + [
+        depset(f[ProtoInfo].transitive_sources)
+        for f in ctx.attr.deps
+    ])
+
+    transitive_externs = depset(transitive = [depset(ctx.attr.extern)] + [
+        f[ProstGenInfo].transitive_externs
+        for f in ctx.attr.extern
+    ])
+
+    extern_descriptors = depset([
+        f[ProstGenInfo].descriptor_set
+        for f in transitive_externs.to_list()
+    ])
+
+    transitive_proto_path = depset(
+        transitive = [
+            f[ProstGenInfo].transitive_proto_path
+            for f in ctx.attr.extern
+        ] + [
+            f[ProtoInfo].transitive_proto_path
+            for f in ctx.attr.deps
+        ],
+    )
+
+    args.add_all(direct_sources)
+    args.add("--output=" + output_dir.path)
+    if ctx.attr.grpc:
+        args.add("--grpc")
+    args.add_all(
+        transitive_externs,
+        map_each = _external_flag,
+    )
+    args.add_all(transitive_proto_path, format_each = "-I%s")
+
+    rustfmt = ctx.toolchains["@io_bazel_rules_rust//rust:toolchain"].rustfmt
+    ctx.actions.run(
+        inputs = depset(
+            transitive = [transitive_sources, extern_descriptors],
+        ),
+        env = {
+            # We've patched rustfmt to expect that it can resolve the path to
```

Comment:
Super minor nit: `s/rustfmt/prost/` ?

Reply:
ahh, yeah.  That should be prost-build, not rustfmt.